pkg/reconciler/timeout_handler_test.go

```diff
@@ -259,3 +259,80 @@ func TestWithNoFunc(t *testing.T) {
 	testHandler.CheckTimeouts()
 
 }
+
+// TestGetTaskRunEndTime checks that the end time calculated for a given taskrun is consistent
+// and correctly uses the default taskrun timeout.
+func TestGetTaskRunEndTime(t *testing.T) {
+	timeNow := time.Now()
+	timeInTwoSeconds := timeNow.Add(2 * time.Second)
+	timeAfterDefaultTimeout := timeNow.Add(defaultTimeout)
+	testCases := []struct {
+		startTime       time.Time
+		optionalTimeout time.Duration
+		expectedEndTime time.Time
+	}{
+		{
+			startTime:       timeNow,
+			optionalTimeout: 2 * time.Second,
+			expectedEndTime: timeInTwoSeconds,
+		},
+		{
+			startTime:       timeNow,
+			expectedEndTime: timeAfterDefaultTimeout,
+		},
+	}
+	for i, tc := range testCases {
+		taskRun := tb.TaskRun("test-taskrun-get-endtime", testNs, tb.TaskRunSpec(
+			tb.TaskRunTaskRef(simpleTask.Name, tb.TaskRefAPIVersion("a1")),
+		), tb.TaskRunStatus(tb.Condition(apis.Condition{
+			Type:   apis.ConditionSucceeded,
+			Status: corev1.ConditionUnknown}),
+			tb.TaskRunStartTime(tc.startTime),
+		))
+		if tc.optionalTimeout != time.Duration(0) {
+			tb.TaskRunTimeout(tc.optionalTimeout)(&taskRun.Spec)
+		}
+		endTime := GetTaskRunEndTime(taskRun)
+		if !endTime.Equal(tc.expectedEndTime) {
+			t.Fatalf("testcase %d expected %q received %q", i+1, tc.expectedEndTime.Format(time.StampNano), endTime.Format(time.StampNano))
+		}
+	}
+}
+
+// TestSetTaskRunTimer checks that the SetTaskRunTimer method correctly calls the TaskRun
+// callback after a set amount of time.
+func TestSetTaskRunTimer(t *testing.T) {
+	taskRun := tb.TaskRun("test-taskrun-arbitrary-timer", testNs, tb.TaskRunSpec(
+		tb.TaskRunTaskRef(simpleTask.Name, tb.TaskRefAPIVersion("a1")),
+		tb.TaskRunTimeout(2*time.Second),
+	), tb.TaskRunStatus(tb.Condition(apis.Condition{
+		Type:   apis.ConditionSucceeded,
+		Status: corev1.ConditionUnknown}),
+		tb.TaskRunStartTime(time.Now().Add(-10*time.Second)),
+	))
+
+	d := test.Data{
+		TaskRuns: []*v1alpha1.TaskRun{taskRun},
+		Tasks:    []*v1alpha1.Task{simpleTask},
+		Namespaces: []*corev1.Namespace{{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testNs,
+			},
+		}},
+	}
+	stopCh := make(chan struct{})
+	c, _ := test.SeedTestData(t, d)
+	observer, _ := observer.New(zap.InfoLevel)
+	testHandler := NewTimeoutHandler(c.Kube, c.Pipeline, stopCh, zap.New(observer).Sugar())
+	dur := 50 * time.Millisecond
+	done := false
+	callback := func(_ interface{}) {
+		done = true
+	}
+	testHandler.SetTaskRunCallbackFunc(callback)
+	testHandler.SetTaskRunTimer(taskRun, dur)
+	<-time.After(100 * time.Millisecond)
```

Comment:
Hmm, I always get a weird feeling when I see hardcoded sleeps in tests. What about another channel you wait on with a select (and a timeout like this) that you write to in the callback?

Reply:
Nice, updated to a `select {}` that relies on a done channel and an expiration timer.